### ISSUE TITLE
Look into replacing PhantomJS with Electron

### ISSUE THREAD

**Issue by StephanBijzitter** [prerender/prerender#318]
Electron is a lot newer and faster, it should give significant performance improvements.

**Comment by thoop** [comment]
Thanks for the suggestion! Looks like Electron is more of a way to build cross platform apps than a headless browser so it might not work for us. Is that true?

**Comment by StephanBijzitter** [comment]
Absolutely not! It's just a chromium based browser that can run with or without head (like a chicken!)

**Comment by thoop** [comment]
It looks like it's not natively headless just yet: https://github.com/electron/electron/issues/228

**Comment by IamCarbonMan** [comment]
Chrome is working on native headless mode, so there's no reason Electron won't be capable of this in the near future. (I believe that was mentioned in [electron/electron#228](https://github.com/electron/electron/issues/228) as well.)

The only thing that really has to happen is to decouple the renderer from the rest of the application enough that the rest of the application can run without having to output graphics somewhere. Once this is accomplished, you have headless mode, which doesn't require any graphical dependencies on the host and doesn't use the resources necessary for rendering.

**Comment by thoop** [comment]
Yeah that would be awesome. We'll definitely try it out once it's ready and see what the performance looks like.

**Comment by rogierslag** [comment]
Any plans on this?

**Comment by thoop** [comment]
No plans on moving to electron but we are testing changes to support headless Chrome once that's ready.

**Comment by hbakhtiyor** [comment]
@thoop since headless chrome already released, any news from your side? will you use headless_shell or just chrome with --headless option?

**Comment by thoop** [comment]
We'll try to use chrome with the --headless option since it looks like they are going to get rid of headless_shell once that command line option is ready for GA.

Headless Chrome has been "released" but we won't officially begin to use it until it's further along: https://bugs.chromium.org/p/chromium/issues/list?q=label:Proj-Headless

**Comment by messense** [comment]
Hello all, we have implemented a Headless Chrome based Prerender in Python with limited features(Only `window.prerenderReady` inspection supported for now):

https://github.com/bosondata/prerender

**Comment by thoop** [comment]
Just a note for future people finding this issue, the above comment is not related to our Prerender server even though it's named similarly. I would not recommend using it if you are trying to use Prerender.io. Just want to prevent any confusion!

**Comment by messense** [comment]
If you are looking for a stable service, I would also recommend to use Prerender.io.

**Comment by Remco75** [comment]
any progress& I use html5 audio features that phantom does not support.... :-( So my pages do not render. Headless Chrome would be the solution!

**Comment by thoop** [comment]
No progress yet, sorry! @Remco75 You might be able to use javascript to check for the html5 audio features and disable them if they aren't available so at least your page renders correctly.

**Comment by Remco75** [comment]
@thoop, thnx, that's exactly what I did, but I really dislike it, the only browser lacking audio is phantom. But at least it works. Would be nice to remove this hacky solution. +1 for headless chrome:-)

**Comment by thoop** [comment]
I agree!

**Comment by GeKorm** [comment]
Worth mentioning that the main PhantomJS maintainer just announced he's stepping down.  
https://groups.google.com/forum/#!topic/phantomjs/9aI5d-LDuNE

**Comment by thoop** [comment]
I saw that. We are very thankful for all of the work they've done, especially Vitaly.

We will continue to make progress with our Prerender server that uses Headless Chrome but won't release it until Headless Chrome is production ready. They still have some issues that prevent it from being production ready: https://bugs.chromium.org/p/chromium/issues/list?q=label:Proj-Headless

Looks like headless firefox is in the works too, so we'll want to keep an eye on that as well: https://bugzilla.mozilla.org/show_bug.cgi?id=1338004

We will continue to use PhantomJS 2.1.1 until Headless Chrome is done.

**Comment by gyzerok** [comment]
@thoop fyi https://github.com/GoogleChrome/puppeteer

**Comment by thoop** [comment]
Thanks! We have our own `chrome` branch going that uses the same underlying interface to Chrome as puppeteer. We've just customized things a little more for our usage so we aren't using puppeteer directly.

We already have lots of Prerender.io users rendering pages in production on our Chrome servers and it's working great so far. We won't close this until we move all of our users to Chrome and make Chrome the main branch.

**Comment by gyzerok** [comment]
@thoop great job!

Do you know how big is perf gain?

Looking forward to be able to use Chrome version of this package! :)

**Comment by thoop** [comment]
The performance (CPU/RAM usage) is actually pretty similar to PhantomJS but the reliability of Chrome over PhantomJS and support for newer features makes Chrome a million times better.

**Comment by hadifarnoud** [comment]
how can I install Chrome for Heroku?

**Comment by thoop** [comment]
I haven't tested this yet, but you might try something like:

https://github.com/heroku/heroku-buildpack-google-chrome

I see that runs Chrome with `--no-sandbox` so you might want to make sure that doesn't cause any security complications. Also, I'm not sure how that might work with our Prerender server since we try to start Chrome on our own.

Going to close this issue since headless Chrome has been merged into master.

**Comment by hadifarnoud** [comment]
I tried the buildpack you mentioned. this is my server.js
```js
#!/usr/bin/env node
var prerender = require('./lib');

var server = prerender();
var server = prerender({
    chromeLocation: '/usr/bin/google-chrome'
});
server.use(prerender.sendPrerenderHeader());
// server.use(prerender.blockResources());
server.use(prerender.removeScriptTags());
server.use(prerender.httpHeaders());
// server.use(prerender.inMemoryHtmlCache());
// server.use(prerender.s3HtmlCache());

// server.use(require('prerender-redis-cache'));
server.start();
```
the app keeps crashing and still says chrome not found

**Comment by thoop** [comment]
@hadifarnoud please create a new issue so we can try to help out.
